Neo4j upgrade to lucene 5

MishaDemianenko opened the pull request:
To be able to use new lucene features and improvements,
cleanup legacy code that was written for old no longer supported 3.6.2 version
we giving a try to migrate neo to latest 5.x available version.

This change set represent only part of required work and includes:
changes done by Paul Horn, fix of failed tests (excluding upgrade testing - those are disabled)

This change does not include cleanups/rethinking/logic update/compatibility code and should only be treated as baseline.

knutwalker committed:
Upgrade Lucene to 5.2.1

knutwalker committed:
Reuse document instances

knutwalker committed:
Optimize search for when only one result is required

knutwalker committed:
Use the proper collector for hits counting

knutwalker committed:
Replace repeated search in Scan Store with continuous DocValues read

knutwalker committed:
Replace usage of Hits with DocValuesCollector

knutwalker committed:
Add tests for DocValuesCollector

knutwalker committed:
Upgrade to Lucene 5.3.0

davidegrohmann committed:
Bump the store version and add migration path from 2.3 to 3.0

MishaDemianenko committed:
Neo4j upgrade to lucene 5.

To be able to use new lucene features and improvements,
cleanup legacy code that was written for old no longer supported 3.6.2 version
we giving a try to migrate neo to latest 5.x available version.

This changeset represent only part of required work and includes:
changes done by Paul Horn, fix of failed tests (excluding upgrade testing - those are disabled)

This change does not include cleanups/rethinking/logic update/compatibility code and should only be treated as baseline.

MishaDemianenko committed:
Restore and upgrade commented out tests, update range seek tests.

Cover range seek queries with a bunch of tests to check inclusive, exlusive values in intervals,
Update todo items.
Keep Double.NaN handling commented - waiting for cypher team to update handling and documentation.

MishaDemianenko committed:
Shuffle lucene-index tests move rules, fields, junit hook methors on top.
Remove IsEmpty, switch to standard matcher instead.
Remove separate process from RecoveryTest.

MishaDemianenko committed:
Remove old tools for rebuilding lucene compound files and segment infos.

Please use org.apache.lucene.index.CheckIndex instead.

MishaDemianenko committed:
Introduce properties for index writer configuration to be able to dynamically adapt them.

MishaDemianenko committed:
Introduce org.neo4j.index.lucene.legacy package for legacy indexes.

Move org.neo4j.index.impl.lucene into org.neo4j.index.lucene.legacy.

MishaDemianenko committed:
Iterate on all fields except id during index sampling

MishaDemianenko committed:
Formatting

MishaDemianenko committed:
Update store migration to remove schema and label indexes as part of 3.0 migration.

Lucene 5.x does not support existent 3.x indexes, so as part of migration we will remove all old indexes, so they will be rebuild.
Restore all ignored upgrage tests.
Current path are only valid for schema and label indexes, legacy indexes will be addressed in a separate commit.

MishaDemianenko committed:
Introduce neo4j upgrader artifact to be able to migrate oblsolete versions of lucene indexes.

MishaDemianenko committed:
Add more license files and update test to ignore directory traversal order

MishaDemianenko committed:
More licenses and notices, update packaging to contain new lucene libs in lib, exclude them from system/lib

MishaDemianenko committed:
Rename upgrade to lucene-index-upgrade, licences files for manual

MishaDemianenko committed:
Update documentation

MishaDemianenko committed:
Update 2.3 store file to include legacy indexes to check legacy index migrations.

MishaDemianenko committed:
Move StoreMigrationTool into separate StoreMigration tool in tools module.
Add possibility to run lucene upgrade tests from IDE.

Tool to migrate store updated to migrate schema and obsolete lucene indexes
and moved from kernel into separate tools module to join other tools.
EmbeddedJarLoaded was updated to provide possibility to load libs from build target/lib directory
to be able to run tests from IDE.

MishaDemianenko committed:
Introduce DatabaseMigrator as concept responsible for whole store migration.

Introduce DatabaseMigrator as holder and builder of StoreUpgrader and all of it participants.
Update progress monitor to be able to monitor whole store migration and be propagated to each migration participant.
Cleanup StoreMigrationParticipant interface from any external service dependencies.
Any dependency that requeired by any particular participant should be passed through constructor to that particular instance.

MishaDemianenko commented:
<b>NB</b>: if https://github.com/neo4j/neo4j/pull/5947 will be merged before this PR - licenses files in new module should be checked before merging

MishaDemianenko commented:
@tinwelint rebased on top of latest and greatest, so can you please start reviewing we want this to be eventually merged, thx a lot

tinwelint commented:
Thanks, I will do just that.

davidegrohmann commented:
@MishaDemianenko needs rebase and note that #5947 has been merged

thobe commented on community/kernel/src/test/java/org/neo4j/graphdb/IndexingAcceptanceTest.java:
```diff
@@ -513,17 +513,17 @@ public void shouldIncludeNodesCreatedInSameTxInIndexSeekByPrefix()
             ReadOperations readOperations = statement.readOperations();
             IndexDescriptor descriptor = indexDescriptor( readOperations, index );
-            found.addAll( readOperations.nodesGetFromIndexRangeSeekByPrefix( descriptor, "Karl" ) );
+            found.addAll( readOperations.nodesGetFromIndexRangeSeekByPrefix( descriptor, "Carl" ) );
```
I'm curious.... does anyone remember why these strings were changed? @knutwalker - do you remember?

knutwalker replied:
@thobe There's no technical reason behind this. I was working on this particular test for a bit and since my nephew is named Carl with a C i felt the need to change this test accordingly.

btw, nice that this is getting worked in after all :+1:

tinwelint commented on community/kernel/src/main/java/org/neo4j/graphdb/index/IndexImplementation.java:
```diff
@@ -35,6 +35,9 @@
 public interface IndexImplementation extends Lifecycle
 {
+
+    File getStoreDirectory( File storeDir );
```
Javadocs please. You're overriding the term "store directory", at least explain why.

tinwelint commented on community/kernel/src/main/java/org/neo4j/helpers/Settings.java:
```diff
@@ -830,6 +832,23 @@ public String apply( Function<String, String> settings )
         };
     }
 
+    public static double getDouble(String propertyName, double defaultValue)
```
I think this fits better in [FeatureToggles](https://github.com/neo4j/neo4j/blob/3.0/community/unsafe/src/main/java/org/neo4j/unsafe/impl/internal/dragons/FeatureToggles.java).

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/DatabaseMigrator.java:
```diff
@@ -0,0 +1,95 @@
+ */
+public class DatabaseMigrator
+{
+    private MigrationProgressMonitor progressMonitor;
```
All these should be final

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreMigrationTool.java:
```diff
@@ -1,130 +0,0 @@
- *
- * @see StoreMigrator
- */
-public class StoreMigrationTool
```
Could this have been a tool used out in the wild and we are supposed to have and now we just deleted?

MishaDemianenko replied:
That is moved to a tools module it's not deleted

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java:
```diff
@@ -333,4 +311,16 @@ public UnexpectedUpgradingStoreVersionException( String filename, String expecte
         }
     }
+
+    public static class DatabaseNotCleanlyShutDown extends UnableToUpgradeException
```
Would be more expected to end with `Exception` in the class name, I had to read it twice.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/legacystore/v21/propertydeduplication/PropertyDeduplicator.java:
```diff
@@ -176,13 +179,16 @@ private void resolveConflicts(
-        try ( IndexLookup indexLookup = new IndexLookup( schemaStore, schemaIndexProvider );
-              IndexedConflictsResolver indexedConflictsResolver =
-                      new IndexedConflictsResolver( duplicateClusters, indexLookup, nodeStore, propertyStore ) )
+        if (isIndexStorageNotEmpty( storeDir ))
```
Please don't have negations in method names and meanings like this. Much better with:

`if ( !isIndexStorageEmpty( storeDir ) )` or similar.

tinwelint commented on community/kernel/src/test/java/org/neo4j/kernel/impl/storemigration/MigrationTestUtils.java:
```diff
@@ -297,4 +310,29 @@ public boolean test( StoreFile item )
+        fileSystem.truncate( logFile, logPosition.getByteOffset() );
+//
+//        // do it again in case there are multiple checkpoints at the end of the tx log
+//        removeCheckPointFromTxLog( fileSystem, workingDirectory );
```
Confusing comment... should this be here or not?

tinwelint commented on community/lucene-index/src/main/java/org/neo4j/index/lucene/legacy/AbstractIndexHits.java:
```diff
@@ -17,7 +17,7 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
-package org.neo4j.index.impl.lucene;
+package org.neo4j.index.lucene.legacy;
```
Shouldn't we keep `impl` in there? We have quite generic fully qualified class name pattern matching where something containing `.impl.` is automatically hidden from javadocs and things like that.

tinwelint commented on community/lucene-index/src/main/java/org/neo4j/index/lucene/legacy/IndexType.java:
```diff
@@ -288,36 +303,60 @@ Query query( String keyOrNull, Object value, QueryContext contextOrNull )
-            field = new Field( key, value.toString(), Store.YES, analyzed );
+            field = new Field( key, value.toString(), fieldType );
+        }
+        // TODO: ensure  omitNorms on fieldType?
```
Should we do anything about this TODO, i.e. about `omitNorms`?

MishaDemianenko replied:
@tinwelint  currently people can create fields with omitNorms false and can have a functionality that depends on it, do no think that we can and to remove it. wdyt?

tinwelint commented on community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/DocValuesCollector.java:
```diff
@@ -0,0 +1,852 @@
+        int size = getTotalHits();
+        if ( size == 0 )
+        {
+            return new AbstractIndexHits<Document>()
```
This should be a static singleton instead of a new instance every time... there should already be something like this somewhere in the code base.

tinwelint commented on community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/DocValuesCollector.java:
```diff
@@ -0,0 +1,852 @@
+        {
+            try
+            {
+                if ( ensureValidDisi() )
```
What is "disi"? Please document that somewhere close by.

MishaDemianenko replied:
that is commonly used abbreviation for DocumentIdSetIterator, will rename it to avid confusion

tinwelint commented on community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/DocValuesCollector.java:
```diff
@@ -0,0 +1,852 @@
+    {
+
+        /** The {@code LeafReaderContext} for this segment. */
+        public final LeafReaderContext context;
```
Are these public by design or by mistake?

tinwelint commented on community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/LuceneIndexAccessor.java:
```diff
@@ -202,18 +201,25 @@ private void closeIndexResources() throws IOException
+        }
+        catch ( IOException e )
+        {
+            throw new RuntimeException( e );
```
Is really RuntimeException an appropriate exception here? We should be explicit with exception handling in as many places as possible.

tinwelint commented on community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/LuceneAllDocumentsReader.java:
```diff
@@ -37,7 +42,14 @@ public LuceneAllDocumentsReader(
+        }
+        catch ( IOException e )
+        {
+            throw new RuntimeException( e );
```
Is RuntimeException the appropriate exception to throw here? IOException or some other checked exception would make things more explicit.

tinwelint commented:
@MishaDemianenko A rebase would be handy. I like these changes in general and the migration code looks understandable and to the point. Will move on to hands-on testing of a couple of things in here...

MishaDemianenko commented:
@tinwelint sure thx, will fix/check comments and rebase; will ping you when done!

tinwelint commented:
@MishaDemianenko ran into a problem right away:

```
Exception in thread "main" java.lang.RuntimeException: Error starting org.neo4j.kernel.impl.factory.CommunityFacadeFactory, C:\Users\Matilas\dev\neo4j\qa\target\test-data\shell-db
    at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.newFacade(GraphDatabaseFacadeFactory.java:143)
    at org.neo4j.kernel.impl.factory.CommunityFacadeFactory.newFacade(CommunityFacadeFactory.java:43)
    at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.newFacade(GraphDatabaseFacadeFactory.java:108)
    at org.neo4j.graphdb.factory.GraphDatabaseFactory.newDatabase(GraphDatabaseFactory.java:129)
    at org.neo4j.graphdb.factory.GraphDatabaseFactory$1.newDatabase(GraphDatabaseFactory.java:117)
    at org.neo4j.graphdb.factory.GraphDatabaseBuilder.newGraphDatabase(GraphDatabaseBuilder.java:185)
    at tooling.StartDbWithShell.main(StartDbWithShell.java:37)
Caused by: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.kernel.NeoStoreDataSource@a0c904b' was successfully initialized, but failed to start. Please see attached cause exception.
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:462)
    at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:111)
    at org.neo4j.kernel.impl.transaction.state.DataSourceManager.start(DataSourceManager.java:98)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:452)
    at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:111)
    at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.newFacade(GraphDatabaseFacadeFactory.java:139)
    ... 6 more
Caused by: java.lang.NullPointerException
    at org.neo4j.io.fs.FileUtils.copyRecursively(FileUtils.java:320)
    at org.neo4j.io.fs.FileUtils.copyRecursively(FileUtils.java:315)
    at org.neo4j.io.fs.DefaultFileSystemAbstraction.copyRecursively(DefaultFileSystemAbstraction.java:175)
    at org.neo4j.kernel.impl.storemigration.participant.LegacyIndexMigrator.moveMigratedFiles(LegacyIndexMigrator.java:98)
    at org.neo4j.kernel.impl.storemigration.StoreUpgrader.moveMigratedFilesToStoreDirectory(StoreUpgrader.java:197)
    at org.neo4j.kernel.impl.storemigration.StoreUpgrader.migrateIfNeeded(StoreUpgrader.java:133)
    at org.neo4j.kernel.impl.storemigration.DatabaseMigrator.migrate(DatabaseMigrator.java:93)
    at org.neo4j.kernel.NeoStoreDataSource.upgradeStore(NeoStoreDataSource.java:570)
    at org.neo4j.kernel.NeoStoreDataSource.start(NeoStoreDataSource.java:450)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:452)
    ... 11 more
```

To reproduce:
- Start up a database on latest 2.3.2-SNAPSHOT (I can imagine any 2.3)
- Shut down
- Start up the (empty) database on this branch w/ `allow_store_upgrade=true`
- Boom

tinwelint commented:
@MishaDemianenko another problem with a populated db:

```
Exception in thread "main" java.lang.RuntimeException: Error starting org.neo4j.kernel.impl.factory.CommunityFacadeFactory, C:\Users\Matilas\Desktop\db-with-legacy-indexes
    at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.newFacade(GraphDatabaseFacadeFactory.java:143)
    at org.neo4j.kernel.impl.factory.CommunityFacadeFactory.newFacade(CommunityFacadeFactory.java:43)
    at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.newFacade(GraphDatabaseFacadeFactory.java:108)
    at org.neo4j.graphdb.factory.GraphDatabaseFactory.newDatabase(GraphDatabaseFactory.java:129)
    at org.neo4j.graphdb.factory.GraphDatabaseFactory$1.newDatabase(GraphDatabaseFactory.java:117)
    at org.neo4j.graphdb.factory.GraphDatabaseBuilder.newGraphDatabase(GraphDatabaseBuilder.java:185)
    at tooling.StartDbWithShell.main(StartDbWithShell.java:37)
Caused by: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.kernel.NeoStoreDataSource@44770933' was successfully initialized, but failed to start. Please see attached cause exception.
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:462)
    at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:111)
    at org.neo4j.kernel.impl.transaction.state.DataSourceManager.start(DataSourceManager.java:98)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:452)
    at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:111)
    at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.newFacade(GraphDatabaseFacadeFactory.java:139)
    ... 6 more
Caused by: org.neo4j.kernel.impl.storemigration.StoreUpgrader$UnableToUpgradeException: Failure cleaning up after migration
    at org.neo4j.kernel.impl.storemigration.StoreUpgrader.cleanup(StoreUpgrader.java:186)
    at org.neo4j.kernel.impl.storemigration.StoreUpgrader.migrateIfNeeded(StoreUpgrader.java:145)
    at org.neo4j.kernel.impl.storemigration.DatabaseMigrator.migrate(DatabaseMigrator.java:93)
    at org.neo4j.kernel.NeoStoreDataSource.upgradeStore(NeoStoreDataSource.java:570)
    at org.neo4j.kernel.NeoStoreDataSource.start(NeoStoreDataSource.java:450)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:452)
    ... 11 more
Caused by: java.nio.file.DirectoryNotEmptyException: C:\Users\Matilas\Desktop\db-with-legacy-indexes\upgrade\index\lucene\node
    at sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:266)
    at sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:103)
    at java.nio.file.Files.delete(Files.java:1126)
    at org.neo4j.io.fs.FileUtils$1.postVisitDirectory(FileUtils.java:91)
    at org.neo4j.io.fs.FileUtils$1.postVisitDirectory(FileUtils.java:1)
    at java.nio.file.Files.walkFileTree(Files.java:2688)
    at java.nio.file.Files.walkFileTree(Files.java:2742)
    at org.neo4j.io.fs.FileUtils.deletePathRecursively(FileUtils.java:75)
    at org.neo4j.io.fs.FileUtils.deleteRecursively(FileUtils.java:70)
    at org.neo4j.io.fs.DefaultFileSystemAbstraction.deleteRecursively(DefaultFileSystemAbstraction.java:133)
    at org.neo4j.kernel.impl.storemigration.participant.StoreMigrator.cleanup(StoreMigrator.java:886)
    at org.neo4j.kernel.impl.storemigration.StoreUpgrader.cleanup(StoreUpgrader.java:181)
    ... 16 more
```

a db that had plenty of legacy indexes in it.

MishaDemianenko committed:
Update core-edge licenses and upgrade configuration handling

MishaDemianenko committed:
Address migration comments

Update org.neo4j.index.lucene.legacy package to org.neo4j.index.impl.lucene.legacy.
Usage of FeatureToggles. Renaming, more descriptive method and variable names.

MishaDemianenko committed:
Address migration comments

Introduce specific exceptions to be thrown in case of IOExceptions during releasing/acquiring lucene IndexSearcher, document retrieval.
Handling of empty source directory during legacy index migration.
Use correct separator for jar resources.

MishaDemianenko commented:
@tinwelint should be fixed and rebased; except last trace, was not able to get it, did check on windows VM; but if you still can reproduce it will be glad to test it on that store and add yet another IT.

tinwelint commented:
Thanks, I'll give both my cases a spin again

tinwelint commented:
First works, second still fails

chrisvest commented:
@tinwelint Can you wrap that case up as a test? Sounds like a good addition.

tinwelint commented:
@chrisvest yes, I already sent that to Misha directly.

MishaDemianenko commented:
@chrisvest yes will add that case as integration test

MishaDemianenko committed:
Fix windows upgrade problem with non ability to delete legacy index directory.

Release index directory stream resources on directory content check.
Move files from upgrade directory to store directory instead of copy.

MishaDemianenko committed:
Add integration test for legacy index migration for cases:
       * when we do not have any indexes
       * when we have node and relationships indexes

davidegrohmann commented:
@MishaDemianenko needs rebase

MishaDemianenko commented:
@tinwelint rebased and green

astesin commented:
Would you please mind telling us, which exact version of Lucene is planned to go into 3.0.0 neo4j release? (stable Lucene itself is of 5.5 now AFAIK). Thanks!
WBR,
Andrii

chrisvest commented:
@astesin Our plan was to migrate to the "latest version 5" which at present means we've upgraded to 5.5.0: https://github.com/neo4j/neo4j/blob/3.0/pom.xml#L28

astesin commented:
Great news, thank you!

On Fri, Mar 4, 2016 at 9:47 AM, Chris Vest notifications@github.com wrote:

> @astesin https://github.com/astesin Our plan was to migrate to the
> "latest version 5" which at present means we've upgraded to 5.5.0:
> https://github.com/neo4j/neo4j/blob/3.0/pom.xml#L28
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/neo4j/neo4j/pull/5861#issuecomment-192167970.
